pkg/splunk/enterprise/util.go

```diff
@@ -232,6 +234,203 @@ func GetSmartstoreRemoteVolumeSecrets(volume enterpriseApi.VolumeSpec, client sp
+	}
+
+	return availDiskSpace, err
+}
+
+// handleAppInstallPerAppSrc kicks off the app install work for all apps in an app source
+func handleAppInstallPerAppSrc(client splcommon.ControllerClient, cr splcommon.MetaObject, numOfWorkers uint64,
+	appSrc string, appDeployInfo *enterpriseApi.AppSrcDeployInfo, s3Response splclient.S3Response, s3ClientMgr S3ClientManager, localPath, scope string) error {
+	var err error
+
+	scopedLog := log.WithName("handleAppInstallPerAppSrc").WithValues("cr", cr.GetName(), "namespace", cr.GetNamespace())
+
+	// get the current available disk space for downloading apps on operator pod
+	availableDiskSpace, err := getAvailableDiskSpace()
+	if err != nil {
+		return err
+	}
+
+	// initialize the workpool
+	workPool := NewAppInstallWorkerPool(numOfWorkers,
+		appDeployInfo.AppDeploymentInfoList, s3ClientMgr, s3Response,
+		availableDiskSpace, localPath, scope)
+
+	// kick-off the actual work
+	scopedLog.Info("Starting the work to install apps..")
+	workPool.Start()
+
+	return err
+}
+
+// handleAppInstall is responsible for installing apps on the splunk pod.This includes following -
+// 1. download of apps from remote storage to operator pod.
+// 2. copy of apps from operator pod to splunk pod.
+// 3. untar the app in correct location on splunk pod.
+// 4. And finally, installing app on the splunk pod.
+// This is how the directory structure will look like on operator pod -
+// /opt/splunk/appframework
+// |--<crname>
+//       |--<cluster>
+//       |   |--appSrc_1
+//       |   |   |--app1.tgz_<hash>
+//       |   |
+//       |   |--appSrc_2
+//       |       |--app2.spl_<hash>
+//       |
+//       |--local
+//       |   |--appSrc_3
+//       |       |--app3.spl_<hash>
+//       |
+//       |--clusterWithPreConfig
+//          |--appSrc_4
+//              |--app4.tgz_<hash>
+//
+func handleAppInstall(client splcommon.ControllerClient, cr splcommon.MetaObject,
+	appDeployContext *enterpriseApi.AppDeploymentContext, remoteObjListingMap map[string]splclient.S3Response, appFrameworkConfig *enterpriseApi.AppFrameworkSpec) error {
+	var err error
+
+	scopedLog := log.WithName("handleAppInstall").WithValues("cr", cr.GetName(), "namespace", cr.GetNamespace())
+
+	var scope string
+	for appSrcName, appDeployInfo := range appDeployContext.AppsSrcDeployStatus {
+		scopedLog.Info("Installing apps for app source", "appSource", appSrcName)
+
+		scope = getAppSrcScope(appFrameworkConfig, appSrcName)
+
+		localPath := filepath.Join(splcommon.AppDownloadVolume, cr.GetName(), scope, appSrcName) + "/"
+
+		// create the sub-directories on the volume for downloading scoped apps
+		err = createAppDownloadDir(localPath)
+		if err != nil {
+			return err
+		}
+
+		// get the app source spec to get the volume configuration
+		var vol enterpriseApi.VolumeSpec
+		appSrc, err := getAppSrcSpec(appFrameworkConfig.AppSources, appSrcName)
+		if err != nil {
+			return err
+		}
```

Comment:
Now that this code is going to be yielding and re-entering. `appFrameworkConfig` may be changing across the reconcile entries. In that case, I guess, we need to handle it more gracefully, than treating it as an error.  Probably, for now, we can add a comment here?
I think this is also a test case we need to cover in our functional testing. CC @pdhanoyasplunk

Reply:
@sgontla Trying to understand your comment better: `getAppSrcSpec` should retrieve the app source for the CR whose reconcile we are in, and if that fails, we should throw an error I would assume. Is there a specific scenario you had in mind when you mentioned handle more gracefully?

Reply:
@smohan-splunk, my comment was for the same CR. Say, if the app framework config changes, while the app installation is already in progress.

Reply:
Interesting point. I wonder if there is an easy way to determine if the `appFrameworkConfig.AppSources` has changed midway of App installations, and hence its ok for `getAppSrcSpec` to fail.